File: services/auth-service/pkg/jwt/jwt_test.go
package jwt

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestGenerateToken(t *testing.T) {
	secret := "test-secret"
	expiry := "1h"

	manager := NewJWTManager(secret, expiry)

	payload := TokenPayload{
		ID:    1,
		Email: "test@example.com",
	}

	token, err := manager.GenerateToken(payload)

	require.NoError(t, err, "Should generate token without error")
	assert.NotEmpty(t, token, "Token should not be empty")
	assert.Greater(t, len(token), 20, "Token should be a substantial string")
}

func TestVerifyToken_ValidToken(t *testing.T) {
	secret := "test-secret"
	expiry := "1h"

	manager := NewJWTManager(secret, expiry)

	originalPayload := TokenPayload{
		ID:    42,
		Email: "user@example.com",
	}

	token, err := manager.GenerateToken(originalPayload)
	require.NoError(t, err)

	verifiedPayload, err := manager.VerifyToken(token)

	require.NoError(t, err, "Should verify valid token")
	assert.Equal(t, originalPayload.ID, verifiedPayload.ID)
	assert.Equal(t, originalPayload.Email, verifiedPayload.Email)
}

func TestVerifyToken_InvalidToken(t *testing.T) {
	secret := "test-secret"
	expiry := "1h"

	manager := NewJWTManager(secret, expiry)

	invalidToken := "invalid.jwt.token"

	_, err := manager.VerifyToken(invalidToken)

	assert.Error(t, err, "Should return error for invalid token")
}

func TestVerifyToken_ExpiredToken(t *testing.T) {
	secret := "test-secret"
	expiry := "1ms" // Very short expiry

	manager := NewJWTManager(secret, expiry)

	payload := TokenPayload{
		ID:    1,
		Email: "test@example.com",
	}

	token, err := manager.GenerateToken(payload)
	require.NoError(t, err)

	// Wait for token to expire
	time.Sleep(10 * time.Millisecond)

	_, err = manager.VerifyToken(token)

	assert.Error(t, err, "Should return error for expired token")
}

func TestVerifyToken_WrongSecret(t *testing.T) {
	manager1 := NewJWTManager("secret1", "1h")
	manager2 := NewJWTManager("secret2", "1h")

	payload := TokenPayload{
		ID:    1,
		Email: "test@example.com",
	}

	token, err := manager1.GenerateToken(payload)
	require.NoError(t, err)

	_, err = manager2.VerifyToken(token)

	assert.Error(t, err, "Should return error when verifying with wrong secret")
}

File: services/auth-service/src/utils/__tests__/jwt.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { generateToken, verifyToken } from '../jwt';

describe('JWT Utils', () => {
  beforeAll(() => {
    process.env.JWT_SECRET = 'test-secret-key';
  });

  it('should generate a valid JWT token', () => {
    const payload = { id: 1, email: 'test@example.com' };
    const token = generateToken(payload);

    expect(token).toBeTruthy();
    expect(typeof token).toBe('string');
    expect(token.split('.')).toHaveLength(3); // JWT has 3 parts
  });

  it('should verify and decode a valid token', () => {
    const payload = { id: 1, email: 'test@example.com' };
    const token = generateToken(payload);

    const decoded = verifyToken(token);

    expect(decoded.id).toBe(payload.id);
    expect(decoded.email).toBe(payload.email);
  });

  it('should throw error for invalid token', () => {
    const invalidToken = 'invalid.token.here';

    expect(() => verifyToken(invalidToken)).toThrow();
  });

  it('should throw error for tampered token', () => {
    const payload = { id: 1, email: 'test@example.com' };
    const token = generateToken(payload);
    const tamperedToken = token.slice(0, -5) + 'xxxxx';

    expect(() => verifyToken(tamperedToken)).toThrow();
  });
});
